apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,944 @@
+use url::Url;
+
+#[cfg(not(test))]
+use crate::error::LicenseError;
+use crate::graphql;
+use crate::layers::ServiceBuilderExt;
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::register_plugin;
+#[cfg(not(test))]
+use crate::services::apollo_graph_reference;
+use crate::services::router;
+use crate::Context;
+
+type SharedDeduplicate = Arc<
+    Deduplicate<Box<dyn Fn(Url) -> DeduplicateFuture<JwkSet> + Send + Sync + 'static>, Url, JwkSet>,
+>;
+
+pub(crate) const AUTHENTICATION_SPAN_NAME: &str = "authentication_plugin";
+
+const DEFAULT_AUTHENTICATION_NETWORK_TIMEOUT: Duration = Duration::from_secs(15);
+
+const DEFAULT_AUTHENTICATION_COOLDOWN: Duration = Duration::from_secs(15);
+
+static COOLDOWN: Lazy<AtomicBool> = Lazy::new(|| AtomicBool::new(false));
+
+static CLIENT: Lazy<Result<Client, BoxError>> = Lazy::new(|| {
+    #[cfg(not(test))]
+    apollo_graph_reference().ok_or(LicenseError::MissingGraphReference)?;
+    Ok(Client::new())
+});
+
+struct AuthenticationPlugin {
+    configuration: JWTConf,
+    jwks: SharedDeduplicate,
+    jwks_url: Url,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct JWTConf {
+    /// Retrieve our JWK Set from here
+    jwks_url: String,
+    /// HTTP header expected to contain JWT
+    #[serde(default = "default_header_name")]
+    header_name: String,
+    /// Header value prefix
+    #[serde(default = "default_header_value_prefix")]
+    header_value_prefix: String,
+    /// JWKS retrieval cooldown
+    #[serde(deserialize_with = "humantime_serde::deserialize", default)]
+    #[schemars(with = "String", default)]
+    cooldown: Option<Duration>,
+}
+
+// This is temporary. It will be removed when the plugin is promoted
+// from experimental.
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct ExperimentalConf {
+    /// The JWT configuration
+    jwt: JWTConf,
+}
+
+// We may support additional authentication mechanisms in future, so all
+// configuration (which is currently JWT specific) is isolated to the
+// JWTConf structure.
+/// Authentication
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct Conf {
+    /// The experimental configuration
+    experimental: ExperimentalConf,
+}
+
+fn default_header_name() -> String {
+    http::header::AUTHORIZATION.to_string()
+}
+
+fn default_header_value_prefix() -> String {
+    "Bearer".to_string()
+}
+
+#[async_trait::async_trait]
+impl Plugin for AuthenticationPlugin {
+    type Config = Conf;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let url: Url = Url::from_str(&init.config.experimental.jwt.jwks_url)?;
+        // This function is expected to return an Optional value, but we'd like to let
+        // users know the various failure conditions. Hence the various clumsy map_err()
+        // scattered through the processing.
+        let getter: Box<dyn Fn(Url) -> DeduplicateFuture<JwkSet> + Send + Sync + 'static> =
```

Comment:
I don't mind using `DeduplicateFuture` since it's the type that the crate expects and saves me from forgetting to specify the future contains an `Option`. I think it requires more thought to use `BoxFuture`.

I also don't mind having closures that don't capture anything from their environment. I know it could be a function, but I think it's just as readable to have a closure here.

I could change both of these, but it does feel as though the changes are purely stylistic.

Reply:
yes, this is purely stylistic (although what I propose would avoid one negligible `Box`), but this is just to transmit the general feeling of "I don't understand what's going on" when reading that code, and proposing ways to make it more obvious

Reply:
Sure. I want the code to be readable, but I guess I was finding it readable already. I'll make the change you suggest and separate it out from `Plugin::new()`